rfd/xxxx-product-metrics.md

```diff
@@ -0,0 +1,97 @@
+Our target audience is security-minded engineer, and the way we collect metrics should reflect their expectations.
+Our metrics and reporting must be built with proper privacy and anonymization protocols in place.
+
+## Metrics Anonymity Invariant
+
+All metrics reported by the Cloud, Self-hosted OSS and Enterprise must be anonymized. As a rule of thumb, if anyone intercepts the metrics stream,
+they should not be able to de-anonymize the emails, or any other personally identifiable information.
+
+To achieve that, all data sent as a part of the metrics stream must be anonymized with SHA256 hash of each sensitive field with a randomly 
+generated prefix identifying cluster id.
+
+For example,
+
+Anonymized user ID: HMAC SHA256 hash of a username with a randomly generated cluster-id.
+Anonymized server ID: HMAC SHA256 hash of a server IP with a randomly generated cluster id.
+
+To make sure customer's of the platform can trust the logic, the anonymization code must be published in OSS and reviewed by independent security auditors.
+
+Current usage reporting code is:
+
+https://github.com/gravitational/teleport/blob/master/lib/services/usagereporter.go
+
+Anonymizer code:
+
+https://github.com/gravitational/teleport/blob/master/lib/utils/anonymizer.go
+
+
+## Classification of Metrics
+
+Teleport will emit two types of product metrics - product usage metrics and user behavior metrics.
+There will be two ways to emit metrics - authenticated metrics and unauthenticated metrics submitted to unauthenticated collection endpoint.
+
+## OSS users promise
+
+We understand that OSS users have strong preferences of freedom, anonymity and privacy. That is why all Teleport's OSS metrics, regardless of their type, will be opt-in.
+
+### User behavior metrics
+
+User behavior metrics help product team to understand most used product features, and help to identify places in the product where users 
+struggle. These metrics are always opt-in. All Teleport clients (web, connect, tsh) must always ask user if they are OK to send anonymized usage data.
+
+Here is an example of Teleport Connect's opt-in usage message:
+
+"Are you OK sending anonymized usage data about Teleport Connect? This will help us to improve product".
+
+Unauthenticated metrics are useful to send data directly from clients, e.g. Teleport Connect.
+To prevent flooding backend with a large number of small requests, events will be batched before sending to a service `prehog`.
+
+Here are some examples of usage event submitted by Teleport connect:
+
+connect.cluster.login
+Successful login to a cluster.
+
+Event properties:
+
+cluster_name: string (anonymized)
+user_name: string (anonymized)
+connector_type: string
+os: string (set once on a user properties)
+arch: string (set once on a user properties) - CPU architecture
+os_version: string (set on a user properties)
+connect_version: string (set on a user properties)
+distinct_id: string
```

Comment:
@gzdunek this makes sense, I think your solution is optimal, I'd just mention that distinct_id is basically a unique client ID generated for each instance of the client.

Reply:
For all current (cloud-only) Teleport-generated events we're using the account id from the license file itself as the `distinct_id`, or the license _name_ when the account id is missing from the license; in both cases, that's a randomly generated UUID. All other user-chosen strings are HMACed with the cluster ID from the `cluster_name` resource.

Should Connect use the same anonymization so we can crossreference the Connect events and the Teleport events, or should Connect explicitly use some other key so that cross-referencing is not possible?